Add media partials for audio and video

markusweigelt opened the pull request:
**What is a 'mediaPartial'?**

- Media partial consist of logicalDivision with a label and a physical division with link to media file, start time and duration

```
<mets:structMap TYPE="PHYSICAL">
    <mets:div ID="PHYS_0000" TYPE="physSequence">
        ...
        <mets:div ID="PHYS_XXXX" ORDER="X" TYPE="track">
            <mets:fptr>
                <mets:area FILEID="FILE_XXXX_DEFAULT" BEGIN="00:00:00" BETYPE="TIME" EXTENT="00:02:30" EXTTYPE="TIME"></mets:area>
            </mets:fptr>
            <mets:fptr FILEID="FILE_XXXX_THUMBS"/>
        </mets:div>
        ...    
    </mets:div>
</mets:structMap>
```

**What is implemented in the PR?**
- Add the `use` with name `mediaPartial` -> divisions marked with this `use` are utilized as media partials
- Divisions with at least one child division with the  `use` `mediaPartial` are active to create media partial for assigned audio and video - only with this condition a button is displayed in media detail (between the media and thumbnail stripe) to create a media partial for audio and video
- When media partials for audio or video are present, the media partials panel is rendered between the media and thumbnail stripe
- With each addition, deletion, or editing of a media partial, they are sorted based on their start time, and the duration until the next media partial or to the duration of media is calculated

**How to?**
- Extend the AV Documentation with documentation for Media Partials
https://github.com/kitodo/kitodo-production/wiki/Verwendung-von-Audio-oder-Video#nutzung-der-medienteile-media-partials

solth commented:
@markusweigelt please rebase against master

solth reviewed (changes requested):
Thanks a lot for this valuable contribution!

This is just a code review. When I find the time, I will test the functionality, too.

markusweigelt commented on Kitodo-DataFormat/src/main/java/org/kitodo/dataformat/access/FileXmlElementAccess.java:
```diff
@@ -144,14 +154,20 @@ DivType toDiv(Map<URI, FileType> mediaFilesToIDFiles,
+                    use.getKey().getMimeType()) && physicalDivision.hasMediaPartialView()) {
+                fptr.setArea(getAreaType(fileId));
+            } else {
+                fptr.setFILEID(fileId);
```
Yes, because it is set as an area attribute.

```
            <mets:fptr>
                <mets:area FILEID="FILE_XXXX_DEFAULT" BEGIN="00:00:00" BETYPE="TIME" EXTENT="00:02:30" EXTTYPE="TIME"></mets:area>
            </mets:fptr>
```

markusweigelt commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java:
```diff
@@ -205,6 +205,16 @@ public boolean isShowingInPreview() {
+     *
+     * @return True if has a media partial
+     */
+    public boolean isMediaPartial() {
```
Funktion is used here https://github.com/kitodo/kitodo-production/blob/1e1a43da8f51311e6ba3bf39859eb27fa7fa470d/Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/partials/media-list-content.xhtml#L19

markusweigelt commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/MediaPartialsPanel.java:
```diff
@@ -0,0 +1,289 @@
+        Collections.sort(logicalDivisions, getLogicalDivisionComparator());
+    }
+
+    private static void generateExtentForMediaPartials(List<LogicalDivision> logicalDivisions, Long mediaDuration) {
```
Added JavaDoc and comments within the method now.

markusweigelt commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/DataEditorForm.java:
```diff
@@ -787,6 +787,9 @@ void setSelectedMedia(List<Pair<PhysicalDivision, LogicalDivision>> media) {
     public boolean isSelected(PhysicalDivision physicalDivision, LogicalDivision logicalDivision) {
         if (Objects.nonNull(physicalDivision) && Objects.nonNull(logicalDivision)) {
+            if (physicalDivision.hasMediaPartialView()) {
+                return selectedMedia.contains(new ImmutablePair<>(physicalDivision, physicalDivision.getLogicalDivisions().get(0)));
```
At the moment, physical divisions with a media partial view are only assigned one logical division. You cannot create a media partial on your own; you need to use the provided function in the user interface. I added a condition to check if there is a logical division. While it ideally should not be necessary, it serves as a safeguard against an IndexOutOfBoundsException in case someone was manipulating the workpiece.

markusweigelt commented:
Implementing review suggestions and adjusted the formatted time to include milliseconds. Refactored the jumping of the audio waveform into its own module to support the selection of milliseconds in videos as well. This module is now rendered only when the media partial is enabled for the surrounding child division of the division where the media is assigned. Changed the behavior of jumping buttons to facilitate jumping between 1, 10, 100 milliseconds, and one second.

matthias-ronge commented on Kitodo-API/src/main/java/org/kitodo/api/dataformat/PhysicalDivision.java:
```diff
@@ -53,6 +53,8 @@ public class PhysicalDivision extends Division<PhysicalDivision> {
      */
     private Map<MediaVariant, URI> mediaFiles = new HashMap<>();
 
+    private MediaPartialView mediaPartialView;
```
Basically it is wrong that the `PhysicalDivision` owns the `MediaPartialView`. **Or rather,** if we change the object model in this way, we should remove the view from the reference from the LogicalDivision altogether, and **only** store it here. I understand what you did and we shouldn't change it now. It would be another major change that should be done separately. But, we have to clean that up again.

markusweigelt replied:
@matthias-ronge I have changed the implementation and removed the inheritance of MediaPartial class from the View.

matthias-ronge replied:
> we should remove the view from the reference from the LogicalDivision altogether

This is something I thought up in the past, but it doesn’t work with the implementation. Please create an issue for this task and assign it to me. I see myself responsible to clean this up.

markusweigelt replied:
I will create a ticket for you as soon as the changes are in the master. Promise!

matthias-ronge reviewed (changes requested):
I have a few suggestions for improvement for you. And then we have a knot in the object model. The previous model was intended to have a logical division with a view referencing a physical division, where the physical division is just the file. You've done it differently now, and I understand why. This is difficult to solve and I wouldn't do it now so as not to jeopardize ongoing development, but we should clean it up afterwards. But this will be a big code change.

Perhaps the best temporary solution would be to not define the `MediaPartialView` as a subclass of `View`. Cleaning up would then mean removing the `View` from its current location, because then it has no function at all.

We can discuss this again separately. I don't want to change everything now, we should take over the development first.

matthias-ronge commented on Kitodo/src/main/java/org/kitodo/production/helper/metadata/MediaPartialHelper.java:
```diff
@@ -0,0 +1,170 @@
+        formattedTime = splittedFormattedTime[0];
+        milliseconds = Integer.parseInt(splittedFormattedTime[1]);
+        String[] time = formattedTime.split(":");
+        return (Integer.parseInt(time[0]) * 3600L + Integer.parseInt(time[1]) * 60L + Integer.parseInt(time[2])) * 1000 + milliseconds;
```
You can use `Duration` for this:
```java
long convertFormattedTimeToMilliseconds(String formattedTime) {
    return Duration.between(LocalTime.MIN, LocalTime.parse(formattedTime)).toMillis();
}
```

markusweigelt commented on Kitodo-API/src/main/java/org/kitodo/api/dataformat/MediaPartialView.java:
```diff
@@ -0,0 +1,71 @@
+
+public class MediaPartialView extends View {
+
+    private String begin;
```
The Duration class, with its toString method, provides a string that does not match the formatting as expected. Therefore, we would need to override or convert the formatting here. That's why I don't see any necessity to change it.

oliver-stoehr reviewed (approved):
I only reviewed the JavaScript files, but they look fine to me!

solth commented on Kitodo/src/main/java/org/kitodo/production/helper/metadata/MediaPartialHelper.java:
```diff
@@ -0,0 +1,148 @@
+        LogicalDivision previousLogicalDivision = null;
+        while (iterator.hasNext()) {
+            LogicalDivision logicalDivision = iterator.next();
+            MediaPartial mediaPartial = logicalDivision.getViews().getFirst().getPhysicalDivision()
```
Don't we have to check whether the logical division _has_ views before retrieving the first? This method is called by a public method, which means anyone could use it in the future and pass a logical division without checking whether it has views or not.

solth commented on Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/partials/media-detail-audio-waveform-tools.xhtml:
```diff
@@ -0,0 +1,26 @@
+<ui:composition
+        xmlns:h="http://xmlns.jcp.org/jsf/html"
+        xmlns:p="http://primefaces.org/ui"
+        xmlns:ui="http://xmlns.jcp.org/jsf/facelets">
```
```suggestion
        xmlns:fn="http://xmlns.jcp.org/jsp/jstl/functions"
        xmlns:ui="http://xmlns.jcp.org/jsf/facelets">
```

solth commented on Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/partials/media-detail-audio-waveform-tools.xhtml:
```diff
@@ -0,0 +1,26 @@
+        xmlns:h="http://xmlns.jcp.org/jsf/html"
+        xmlns:p="http://primefaces.org/ui"
+        xmlns:ui="http://xmlns.jcp.org/jsf/facelets">
+    <p:outputPanel id="audioWaveformTools" rendered="#{fn:startsWith(selectedGalleryMediaContent.mediaViewMimeType, 'audio') and DataEditorForm.galleryPanel.isAudioMediaViewWaveform()}">
```
```suggestion
    <!--@elvariable id="selectedGalleryMediaContent" type="org.kitodo.production.forms.dataeditor.GalleryMediaContent"-->
    <p:outputPanel id="audioWaveformTools" rendered="#{fn:startsWith(selectedGalleryMediaContent.mediaViewMimeType, 'audio') and DataEditorForm.galleryPanel.isAudioMediaViewWaveform()}">
```

solth commented on Kitodo/src/main/resources/messages/messages_de.properties:
```diff
@@ -656,6 +655,14 @@ insert=Einf\u00FCgen
+jumpBackwardOneMillisecond=1 Millisekunde zur\u00FCckspringen
+jumpForwardOneMillisecond=1 Millisekunde vorspringen
+jumpForwardTenMilliseconds=10 Millisekunden vorspringen
+jumpForwardOnHundredMilliseconds=100 Millisekunden vorspringen
```
Could it be this is supposed to be "On**e**Hundred" instead of "OnHundred"?

solth commented on Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/logicalStructure.xhtml:
```diff
@@ -232,6 +240,20 @@
                                 metadataAccordion:logicalMetadataWrapperPanel
                                 galleryWrapperPanel"/>
+            <p:menuitem value="#{msgs.deleteMediaPartial}"
+                        rendered="#{SecurityAccessController.hasAuthorityToDeleteMedia() and DataEditorForm.structurePanel.selectedLogicalNode.type eq StructurePanel.MEDIA_PARTIAL_NODE_TYPE}"
```
```suggestion
                        rendered="#{SecurityAccessController.hasAuthorityToDeleteMedia() and DataEditorForm.structurePanel.selectedLogicalNode.type eq DataEditorForm.structurePanel.MEDIA_PARTIAL_NODE_TYPE}"
```
The class `StructurePanel` is not visible here, so I think we need to retrieve it via the `DataEditorForm`, don't we?  (same above)

markusweigelt replied:
I get following exceptions when i change that to DataEditorForm.structurePanel. 

![image](https://github.com/kitodo/kitodo-production/assets/3832618/dc8201d6-83a2-4649-ae82-cf3ed55ee241)

In this file, other calls to class constants to structure panel are done in the same way as I do. 
https://github.com/kitodo/kitodo-production/blob/2960f1fd0f425170fb6ac5c7fb544c1c6e626955/Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/logicalStructure.xhtml#L115

https://github.com/kitodo/kitodo-production/blob/2960f1fd0f425170fb6ac5c7fb544c1c6e626955/Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/logicalStructure.xhtml#L133

solth commented on Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/physicalStructure.xhtml:
```diff
@@ -81,6 +83,13 @@
                         a:data-order="#{physicalNode.order}"/>
                 <h:outputText value=" 🔗" rendered="#{physicalNode.linked}"/>
             </p:treeNode>
+            <p:treeNode type="#{StructurePanel.MEDIA_PARTIAL_NODE_TYPE}"
```
See above (`DataEditorForm.structurePanel` instead of `StructurePanel.`)

solth commented on Kitodo/src/test/java/org/kitodo/production/forms/dataeditor/MediaPartialFormTest.java:
```diff
@@ -0,0 +1,187 @@
+        // overwrite to test save function
+        LinkedList<PhysicalDivision> ancestorsOfPhysicalDivision = new LinkedList<>();
+        ancestorsOfPhysicalDivision.add(new PhysicalDivision());
+        MockedStatic<MetadataEditor> metadataEditorMockedStatic = Mockito.mockStatic(MetadataEditor.class);
```
This should probably use try-with-resources

solth commented on Kitodo/src/test/java/org/kitodo/production/forms/dataeditor/MediaPartialFormTest.java:
Test should use `assert` methods from `junit.jupiter.Assert` instead of `org.junit.Assert`.

solth commented on Kitodo/src/test/java/org/kitodo/production/forms/dataeditor/MediaPartialPanelTest.java:
Test should use `assert` methods from `junit.jupiter.Assert` instead of `org.junit.Assert`.

solth reviewed (commented):
Thanks a lot again for this pull request. This adds a very useful feature and great value to the application in my opinion.

I have a few small comments and still struggle to build the project with this branch, as I already told you, but I hope I will be able to get it to run and try it shortly.

matthias-ronge commented:
Hello @solth, I can't yet understand the functionality required here. I will explore the theory in more detail

matthias-ronge reviewed (approved):
I don't feel 100% sure that this; but since I don't see any errors, I agree to the pull request. It's a great piece of work and a big step forward for us.

matthias-ronge commented:
Deeper testing revealed: you need to type in the time as 00:00:00.000. This in inconvenient. The dialog should support easier input like 0:22 which it should understand as minute:second. It is not necessary to fix this in the current pull request, but it should be improved some next time.

![Zeit wird nicht erkannt](https://github.com/kitodo/kitodo-production/assets/3040657/4d231381-2001-4493-8739-625cd1b34b7a)

solth commented:
Thanks for implementing the change requests and for the extensive documentation. Could you also add a section for `use="mediaPartial"` to the [general ruleset documentation](https://github.com/kitodo/kitodo-production/wiki/Regelsatz-3.x#einteilungen-mit-besonderer-verwendung)?

markusweigelt commented:
@solth Added to the [general ruleset documentation](https://github.com/kitodo/kitodo-production/wiki/Regelsatz-3.x#schlüssel-mit-besonderer-verwendung)

solth commented:
Thanks for implementing the change requests. I tested the branch and it works well. I just have one more problem with the tests while building the branch:
```
[ERROR] org.kitodo.production.forms.dataeditor.MediaPartialFormTest.testSave -- Time elapsed: 0.854 s <<< ERROR!
java.util.NoSuchElementException
	at java.base/java.util.LinkedList.getLast(LinkedList.java:261)
	at org.kitodo.production.helper.metadata.MediaPartialHelper.addMediaPartialToMediaSelection(MediaPartialHelper.java:148)
	at org.kitodo.production.forms.dataeditor.MediaPartialForm.save(MediaPartialForm.java:97)
	at org.kitodo.production.forms.dataeditor.MediaPartialForm.save(MediaPartialForm.java:89)
	at org.kitodo.production.forms.dataeditor.MediaPartialFormTest.testSave(MediaPartialFormTest.java:117)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:56)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:184)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:148)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:122)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
```
Do you have an idea how to resolve this?

solth commented:
@markusweigelt thanks for fixing the error!

solth commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/MediaPartialForm.java:
```diff
@@ -0,0 +1,190 @@
+import static org.kitodo.production.helper.metadata.MediaPartialHelper.convertFormattedTimeToMilliseconds;
+
+import java.io.Serializable;
+import java.util.List;
```
```suggestion
```
This is not required anymore, it seems.

solth commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/MediaPartialForm.java:
```diff
@@ -0,0 +1,190 @@
+        }
+        if (!isEditable() || (isEditable() && !mediaPartialDivision.getValue().getBegin().equals(getBegin()))) {
+            boolean exists = getMediaSelection().getValue().getChildren().stream().anyMatch(
+                    logicalDivision -> logicalDivision.getViews().size() > 0 && logicalDivision.getViews().getFirst()
```
```suggestion
                    logicalDivision -> !logicalDivision.getViews().isEmpty() && logicalDivision.getViews().getFirst()
```

solth commented on Kitodo/src/main/java/org/kitodo/production/helper/metadata/MediaPartialHelper.java:
```diff
@@ -0,0 +1,182 @@
+        LogicalDivision previousLogicalDivision = null;
+        while (iterator.hasNext()) {
+            LogicalDivision logicalDivision = iterator.next();
+            if (logicalDivision.getViews().size() == 0) {
```
```suggestion
            if (logicalDivision.getViews().isEmpty()) {
```

solth commented on Kitodo/src/main/java/org/kitodo/production/helper/metadata/MediaPartialHelper.java:
```diff
@@ -0,0 +1,182 @@
+
+        LinkedList<PhysicalDivision> physicalDivisions = MetadataEditor.getAncestorsOfPhysicalDivision(
+                mediaSelection.getKey(), workpiece.getPhysicalStructure());
+        if (physicalDivisions.size() > 0) {
```
```suggestion
        if (!physicalDivisions.isEmpty()) {
```

solth commented on Kitodo/src/main/java/org/kitodo/production/helper/metadata/MediaPartialLogicalDivisionComparator.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public int compare(LogicalDivision logicalDivisionA, LogicalDivision logicalDivisionB) {
+        if (logicalDivisionA.getViews().size() > 0 && logicalDivisionB.getViews().size() > 0) {
```
```suggestion
        if (!logicalDivisionA.getViews().isEmpty() && !logicalDivisionB.getViews().isEmpty()) {
```

solth commented on Kitodo/src/main/resources/messages/messages_de.properties:
```diff
@@ -656,6 +655,14 @@ insert=Einf\u00FCgen
 insertAsUncounted=Als unnummeriert einf\u00FCgen
 insertHere=hier einf\u00FCgen
 isInUse=befindet sich in Bearbeitung durch
+jumpBackwardOnSecond=1 Sekunde zur\u00FCckspringen
```
Shouldn't this be `jumpBackwardOneSecond`?

solth commented on Kitodo/src/main/resources/messages/messages_de.properties:
```diff
@@ -656,6 +655,14 @@ insert=Einf\u00FCgen
 insertHere=hier einf\u00FCgen
 isInUse=befindet sich in Bearbeitung durch
+jumpBackwardOnSecond=1 Sekunde zur\u00FCckspringen
+jumpBackwardOnHundredMilliseconds=100 Millisekunden zur\u00FCckspringen
```
Shouldn't this be `jumpBackwardOneHundredMilliseconds`?

solth commented on Kitodo/src/main/resources/messages/messages_en.properties:
```diff
@@ -657,6 +656,14 @@ insert=Insert
 insertAsUncounted=Insert as uncounted
 insertHere=insert here
 isInUse=is in use of
+jumpBackwardOnSecond=Jump backward 1 second
+jumpBackwardOnHundredMilliseconds=Jump backward 100 milliseconds
```
See above

solth reviewed (commented):
Just some very minor remarks!
